src/Microsoft.AspNet.Hosting.Abstractions/IApplicationLifetime.cs

```diff
@@ -11,6 +12,12 @@ namespace Microsoft.AspNet.Hosting
     public interface IApplicationLifetime
     {
         /// <summary>
+        /// Triggered when the application host has fully started and is about to wait
+        /// for a graceful shutdown.
+        /// </summary>
+        event EventHandler ApplicationStarted;
```

Comment:
The benefit of cancellation tokens is that you can both poll it an handle an event.